crates/iota-graphql-rpc/src/types/subscription/filter.rs

```diff
@@ -54,22 +51,44 @@ pub(crate) enum SubscriptionTransactionFilter {
     Function(FqNameFilter),
 }
 
-impl From<SubscriptionTransactionFilter> for StreamTransactionFilter {
-    fn from(value: SubscriptionTransactionFilter) -> Self {
+impl Filter<StoredTransaction> for SubscriptionTransactionFilter {
+    fn matches(&self, transaction: &StoredTransaction) -> bool {
         use SubscriptionTransactionFilter::*;
-        match value {
-            Kind(kind) => StreamTransactionFilter::Kind(kind.into()),
-            SigningAddress(address) => StreamTransactionFilter::SigningAddress(address.into()),
+        match self {
+            Kind(kind) => transaction.transaction_kind == IotaTransactionKind::from(kind) as i16,
+            SigningAddress(address) => transaction
+                .try_into_sender_signed_data()
+                .map(|data| data.transaction_data().sender() == (*address).into())
+                .unwrap_or_default(),
             Function(name) => {
                 let (package, module) = name.into();
-                StreamTransactionFilter::Function { package, module }
+
+                transaction
+                    .try_into_sender_signed_data()
+                    .map(|data| {
+                        data.transaction_data()
+                            .move_calls()
+                            .iter()
+                            .any(|(p, m, f)| match module.as_ref() {
+                                Some(module_function) => {
+                                    let Some(ref function) = module_function.function_name else {
+                                        return **p == package && *m == module_function.module_name;
+                                    };
+                                    **p == package
+                                        && *m == module_function.module_name
+                                        && f == function
+                                }
+                                None => **p == package,
```

Comment:
have you considered to extract this logic to a separate function?

Reply:
fixed: https://github.com/iotaledger/iota/pull/8949/commits/c8cb0612c0941183a241ca92c0c45068d34f2d30